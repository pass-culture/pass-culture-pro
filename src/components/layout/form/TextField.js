/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import {
  composeValidators,
  config,
  createParseNumberValue,
  createValidateRequiredField,
} from 'react-final-form-utils'

import FieldError from './FieldError'

const validateRequiredField = createValidateRequiredField(
  config.DEFAULT_REQUIRED_ERROR
)

export const TextField = ({
  autoComplete,
  className,
  disabled,
  format,
  id,
  label,
  name,
  parse,
  placeholder,
  readOnly,
  renderInner,
  renderValue,
  required,
  title,
  type,
  validate,
  ...inputProps
}) => {
  const requiredValidate =
    required && typeof required === 'function'
      ? required
      : (required && validateRequiredField) || undefined

  return (
    <Field
      format={format}
      name={name}
      validate={composeValidators(validate, requiredValidate)}
      parse={parse || createParseNumberValue(type)}
      render={({ input, meta }) => {
        let value = input.value
        if (!readOnly && typeof value === 'string') {
          value = ''
        }
        return (
          <div
            className={classnames('field text-field', className, {
              'is-read-only': readOnly,
            })}
            id={id}>
            <label
              htmlFor={name}
              className={classnames('field-label', { empty: !label })}>
              {label && (
                <span>
                  <span>{label}</span>
                  {required && !readOnly && (
                    <span className="field-asterisk">*</span>
                  )}
                </span>
              )}
            </label>
            <div className="field-control">
              <div className="field-value flex-columns items-center">
                <div className="field-inner flex-columns items-center">
                  <input
                    {...input}
                    {...inputProps}
                    autoComplete={autoComplete ? 'on' : 'off'}
                    className={`field-input field-${type}`}
                    disabled={disabled || readOnly}
                    placeholder={readOnly ? '' : placeholder}
                    readOnly={readOnly}
                    required={!!required} // cast to boolean
                    title={title}
                    type={readOnly ? 'text' : type}
                    value={value}
                  />
                  {renderInner()}
                </div>
                {renderValue()}
              </div>
              <FieldError meta={meta} />
            </div>
          </div>
        )
      }}
    />
  )
}

TextField.defaultProps = {
  autoComplete: false,
  className: '',
  disabled: false,
  format: null,
  id: null,
  label: '',
  parse: null,
  placeholder: 'Please enter a value',
  readOnly: false,
  renderInner: () => null,
  renderValue: () => null,
  required: false,
  title: null,
  type: 'text',
  validate: null,
}

TextField.propTypes = {
  autoComplete: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  format: PropTypes.func,
  id: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  title: PropTypes.string,
  type: PropTypes.string,
  validate: PropTypes.func,
}

export default TextField
