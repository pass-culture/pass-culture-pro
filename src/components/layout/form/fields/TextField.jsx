import classnames from 'classnames'
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import { composeValidators, createParseNumberValue } from 'react-final-form-utils'

import FieldErrors from '../FieldErrors'
import getRequiredValidate from '../utils/getRequiredValidate'

function getInputValue(inputType, value) {
  const isStringifiedNumber = inputType === 'number' && typeof value === 'string'

  if (isStringifiedNumber) {
    return ''
  }

  return value
}

class TextField extends PureComponent {
  componentDidMount() {
    const { type, isDecimal } = this.props
    if (type === 'number') {
      this.preventEnteringInvalidChars(isDecimal)
    }
  }

  componentWillUnmount() {
    if (this.keypressListener) {
      this.inputElement.removeEventListener(this.keypressListener)
    }
  }

  preventEnteringInvalidChars = (isDecimal) => {
    this.keypressListener = this.inputElement.addEventListener('keypress', event => {
      const validCharacters = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0']

      if (isDecimal) {
        validCharacters.push(',', '.')
      }

      if (!validCharacters.includes(event.key)) {
        event.preventDefault()
      }
    })
  }

  handleRef = _e => {
    this.inputElement = _e
  }

  renderField = ({ input, meta }) => {
    const {
      className,
      disabled,
      id,
      innerClassName,
      isDecimal,
      label,
      min,
      name,
      placeholder,
      readOnly,
      renderInner,
      renderValue,
      required,
      type,
      ...inputProps
    } = this.props
    const inputType = readOnly ? 'text' : type
    const inputValue = getInputValue(inputType, input.value)

    return (
      <div
        className={classnames('field text-field', className, {
          'is-label-aligned': label,
          'is-read-only': readOnly,
        })}
        id={id}
      >
        {label && (
          <label
            className={classnames('field-label')}
            htmlFor={name}
          >
            <span>
              <span>
                {label}
              </span>
              {required && !readOnly && (
                <span className="field-asterisk">
                  {'*'}
                </span>
              )}
            </span>
          </label>
        )}
        <div className="field-control">
          <div className="field-value flex-columns items-center">
            <div className={classnames('field-inner flex-columns items-center', innerClassName)}>
              <input
                id={name}
                {...input}
                {...inputProps}
                className={`field-input field-${type}`}
                disabled={disabled || readOnly}
                min={min}
                placeholder={readOnly ? '' : placeholder}
                readOnly={readOnly}
                ref={this.handleRef}
                required={!!required}
                type={inputType}
                value={inputValue}
              />
              {renderInner()}
            </div>
            {renderValue()}
          </div>
          <FieldErrors meta={meta} />
        </div>
        <div />
      </div>
    )
  }

  render() {
    const { format, name, parse, required, type, validate } = this.props

    return (
      <Field
        format={format}
        name={name}
        parse={parse || createParseNumberValue(type)}
        render={this.renderField}
        validate={composeValidators(validate, getRequiredValidate(required, type))}
      />
    )
  }
}

TextField.defaultProps = {
  className: '',
  disabled: false,
  format: null,
  id: null,
  innerClassName: null,
  isDecimal: true,
  label: '',
  parse: null,
  placeholder: '',
  readOnly: false,
  renderInner: () => null,
  renderValue: () => null,
  required: false,
  type: 'text',
  validate: null,
}

TextField.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  format: PropTypes.func,
  id: PropTypes.string,
  innerClassName: PropTypes.string,
  isDecimal: PropTypes.bool,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  type: PropTypes.string,
  validate: PropTypes.func,
}

export default TextField
