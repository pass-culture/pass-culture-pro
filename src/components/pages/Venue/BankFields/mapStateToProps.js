import { selectCurrentUser } from 'with-login'

import selectUserOffererByOffererIdAndUserIdAndRightsType from 'selectors/selectUserOffererByOffererIdAndUserIdAndRightsType'

const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const {
    params: { offererId },
  } = match
  const currentUser = selectCurrentUser(state)
  const { id: currentUserId } = currentUser || {}

  return {
    adminUserOfferer: selectUserOffererByOffererIdAndUserIdAndRightsType(
      state,
      offererId,
      currentUserId,
      'admin'
    ),
  }
}

export default mapStateToProps
