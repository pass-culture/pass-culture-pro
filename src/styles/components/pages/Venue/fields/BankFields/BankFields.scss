.bank-instructions-label {
  color: $primary;
  font-size: rem(12px);
  font-weight: $weight-bold;
  line-height: normal;
  margin-bottom: rem(18px);
}
