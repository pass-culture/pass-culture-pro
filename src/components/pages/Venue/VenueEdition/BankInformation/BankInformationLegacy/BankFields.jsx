import PropTypes from 'prop-types'
import React from 'react'

import TextField from '../../../../../layout/form/fields/TextField'

const BankFields = ({ adminUserOfferer, areBankInformationProvided, readOnly }) => {
  const areBankInfosReadOnly = readOnly || !adminUserOfferer

  return (
    <div className="section">
      <h2 className="main-list-title">
        {'Informations bancaires'}
        <span className="is-pulled-right fs13 has-text-grey">
          {!adminUserOfferer &&
            "Vous avez besoin d'être administrateur de la structure pour éditer ces informations."}
        </span>
      </h2>
      {!areBankInformationProvided && (
        <p className="bank-instructions-label fs13">
          {
            'Le pass Culture vous contactera prochainement afin d’enregistrer vos coordonnées bancaires. Une fois votre BIC / IBAN renseigné, ces informations apparaitront ci-dessous.'
          }
        </p>
      )}
      <div className="field-group">
        <TextField label="BIC : " name="bic" readOnly={areBankInfosReadOnly} />
        <TextField label="IBAN : " name="iban" readOnly={areBankInfosReadOnly} />
      </div>
    </div>
  )
}

BankFields.defaultProps = {
  adminUserOfferer: false,
  areBankInformationProvided: false,
  readOnly: true,
}

BankFields.propTypes = {
  adminUserOfferer: PropTypes.bool,
  areBankInformationProvided: PropTypes.bool,
  readOnly: PropTypes.bool,
}

export default BankFields
