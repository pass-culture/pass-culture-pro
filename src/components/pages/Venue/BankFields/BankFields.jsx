import PropTypes from 'prop-types'
import React from 'react'

import { TextField } from 'components/layout/form/fields'

const BankFields = ({
  adminUserOfferer,
  initialIban,
  initialThumbCount,
  readOnly,
}) => {
  const areBankInfosReadOnly = readOnly || !adminUserOfferer

  const hasAlreadyRibUploaded = initialIban && initialThumbCount

  return (
    <div className="section">
      <h2 className="main-list-title">
        INFORMATIONS BANCAIRES
        <span className="is-pulled-right is-size-7 has-text-grey">
          {!adminUserOfferer &&
            "Vous avez besoin d'Ãªtre administrateur de la structure pour editer ces informations."}
        </span>
      </h2>
      <div className="field-group">
        <TextField label="BIC : " name="bic" readOnly={areBankInfosReadOnly} />
        <TextField
          label="IBAN : "
          name="iban"
          readOnly={areBankInfosReadOnly}
        />
        {false && (
          <TextField
            label="Justificatif"
            name="rib"
            readOnly={readOnly || areBankInfosReadOnly}
            uploaded={hasAlreadyRibUploaded}
            type="file"
          />
        )}
      </div>
    </div>
  )
}

BankFields.defaultProps = {
  initialIban: null,
  initialThumbCount: null,
  readOnly: true,
}

BankFields.propTypes = {
  initialIban: PropTypes.string,
  initialThumbCount: PropTypes.number,
  readOnly: PropTypes.bool,
}

export default BankFields
