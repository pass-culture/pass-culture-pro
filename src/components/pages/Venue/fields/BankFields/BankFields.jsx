import PropTypes from 'prop-types'
import React from 'react'

import { TextField } from 'components/layout/form/fields'

const BankFields = ({
  adminUserOfferer,
  readOnly,
}) => {
  const areBankInfosReadOnly = readOnly || !adminUserOfferer

  return (
    <div className="section">
      <h2 className="main-list-title">
        INFORMATIONS BANCAIRES
        <span className="is-pulled-right is-size-7 has-text-grey">
          {!adminUserOfferer &&
            "Vous avez besoin d'être administrateur de la structure pour éditer ces informations."}
        </span>
      </h2>
      <div className="field-group">
        <TextField
          label="BIC : "
          name="bic"
          readOnly={areBankInfosReadOnly}
        />
        <TextField
          label="IBAN : "
          name="iban"
          readOnly={areBankInfosReadOnly}
        />
      </div>
    </div>
  )
}

BankFields.defaultProps = {
  adminUserOfferer: false,
  readOnly: true,
}

BankFields.propTypes = {
  adminUserOfferer: PropTypes.bool,
  readOnly: PropTypes.bool,
}

export default BankFields
