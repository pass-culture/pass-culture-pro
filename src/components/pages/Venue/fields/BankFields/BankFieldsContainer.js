import { connect } from 'react-redux'
import { compose } from 'redux'
import { selectCurrentUser } from 'with-login'

import BankFields from './BankFields'
import withFrenchQueryRouter from '../../../../hocs/withFrenchQueryRouter'
import selectUserOffererByOffererIdAndUserIdAndRightsType from '../../../../../selectors/selectUserOffererByOffererIdAndUserIdAndRightsType'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const {
    params: { offererId },
  } = match
  const currentUser = selectCurrentUser(state)
  const { id: currentUserId } = currentUser || {}

  const adminUserOfferer = selectUserOffererByOffererIdAndUserIdAndRightsType(
    state,
    offererId,
    currentUserId,
    'admin'
  )

  return {
    adminUserOfferer: adminUserOfferer,
  }
}

export default compose(
  withFrenchQueryRouter,
  connect(mapStateToProps)
)(BankFields)
