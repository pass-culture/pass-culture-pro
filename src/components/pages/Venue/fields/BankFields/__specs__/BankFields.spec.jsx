import { shallow } from 'enzyme'
import React from 'react'

import BankFields from '../BankFields'
import TextField from '../../../../../layout/form/fields/TextField'

describe('src | components | pages | Venue | fields | BankFields', () => {
  let props

  beforeEach(() => {
    props = {
      adminUserOfferer: true,
      areBankInformationProvided: false,
      readOnly: true,
    }
  })

  it('should match the snapshot', () => {
    // when
    const wrapper = shallow(<BankFields {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('render', () => {
    it('should display information label when user is not admin of the offerer', () => {
      // given
      props.adminUserOfferer = false

      // when
      const wrapper = shallow(<BankFields {...props} />)

      // then
      const title = wrapper.find('h2 > span')
      expect(title.text()).toBe(
        "Vous avez besoin d'être administrateur de la structure pour éditer ces informations."
      )
    })

    it('should not display information label when user is admin of the offerer', () => {
      // given
      props.adminUserOfferer = true

      // when
      const wrapper = shallow(<BankFields {...props} />)

      // then
      const title = wrapper.find('h2 > span')
      expect(title.text()).toBe('')
    })

    it('should display two TextField components with the right props', () => {
      // when
      const wrapper = shallow(<BankFields {...props} />)

      // then
      const textFields = wrapper.find(TextField)
      expect(textFields).toHaveLength(2)
      expect(textFields.at(0).prop('label')).toBe('BIC : ')
      expect(textFields.at(0).prop('name')).toBe('bic')
      expect(textFields.at(0).prop('readOnly')).toBe(true)
      expect(textFields.at(1).prop('label')).toBe('IBAN : ')
      expect(textFields.at(1).prop('name')).toBe('iban')
      expect(textFields.at(1).prop('readOnly')).toBe(true)
    })

    it('should render a label element with text containing instructions for offerer when bic/iban are missing', () => {
      // given
      props.areBankInformationProvided = false

      // when
      const wrapper = shallow(<BankFields {...props} />)

      // then
      const labelElement = wrapper.find('p')
      expect(labelElement.prop('className')).toBe('bank-instructions-label fs13')
      expect(labelElement.text()).toBe(
        'Le pass Culture vous contactera prochainement afin d’enregistrer vos coordonnées bancaires. Une fois votre BIC / IBAN renseigné, ces informations apparaitront ci-dessous.'
      )
    })

    it('should not render a label element with text containing instructions for offerer when bic/iban are provided', () => {
      // given
      props.areBankInformationProvided = true

      // when
      const wrapper = shallow(<BankFields {...props} />)

      // then
      const labelElement = wrapper.find('p')
      expect(labelElement).toHaveLength(0)
    })
  })
})
