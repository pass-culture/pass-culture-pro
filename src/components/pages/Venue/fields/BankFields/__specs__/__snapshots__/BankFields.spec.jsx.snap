// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Venue | fields | BankFields should match the snapshot 1`] = `ShallowWrapper {}`;
